iree/compiler/Dialect/HAL/Transforms/LinkExecutables.cpp

```diff
@@ -47,6 +47,17 @@ class LinkExecutablesPass
         return signalPassFailure();
       }
     }
+
+    // Backends may move target ops from executables into linked executables.
```

Comment:
The `iree/samples/simple_embedding:simple_embedding_test` failure is interesting. The [logs](https://source.cloud.google.com/results/invocations/ac6c8355-88de-4ee7-bb2b-106060fe2739/targets/iree%2Fgcp_ubuntu%2Fbazel%2Flinux%2Fx86-swiftshader%2Fcore%2Fpresubmit/log) are surface-level, but I've narrowed it down under a debugger.

[ExecutableCachePrepare](https://github.com/google/iree/blob/efd78a0b47a46457a644f43d98617d3e279b2a79/iree/modules/hal/hal_module.cc#L671-L681) is being called with null `executable_data`. I looked through the IR to see where `hal.executable_cache.prepare` is used.

Here's the relevant IR from the final translation in `vm.func @__init()` before:

```mlir
    %_simple_mul_ex_dispatch_0_binary_vmla = vm.const.ref.rodata @_simple_mul_ex_dispatch_0_binary_vmla : !vm.ref<!iree.byte_buffer>
    %_simple_mul_ex_dispatch_0_binary_spirv = vm.const.ref.rodata @_simple_mul_ex_dispatch_0_binary_spirv : !vm.ref<!iree.byte_buffer>
    %ref_7 = vm.switch.ref %2[%_simple_mul_ex_dispatch_0_binary_vmla, %_simple_mul_ex_dispatch_0_binary_spirv] else %null : !vm.ref<!iree.byte_buffer>
    %ref_8 = vm.call @hal.executable_cache.prepare(%ref_6, %_executable_layout_0, %c7, %ref_7) : (!vm.ref<!hal.executable_cache>, !vm.ref<!hal.executable_layout>, i32, !vm.ref<!iree.byte_buffer>) -> !vm.ref<!hal.executable>
```

And here's the IR after these linking changes:

```mlir
    %_linked_vmla_binary_vmla = vm.const.ref.rodata @_linked_vmla_binary_vmla : !vm.ref<!iree.byte_buffer>
    %ref_7 = vm.switch.ref %2[%_linked_vmla_binary_vmla] else %null : !vm.ref<!iree.byte_buffer>
    %ref_8 = vm.call @hal.executable_cache.prepare(%ref_6, %_executable_layout_0, %c7, %ref_7) : (!vm.ref<!hal.executable_cache>, !vm.ref<!hal.executable_layout>, i32, !vm.ref<!iree.byte_buffer>) -> !vm.ref<!hal.executable>
    vm.global.store.ref %ref_8, @_executable_linked_vmla : !vm.ref<!hal.executable>
    %_executable_layout_0_9 = vm.global.load.ref @_executable_layout_0 : !vm.ref<!hal.executable_layout>
    %3 = vm.call.variadic @hal.executable_cache.select_format(%ref_6, [%c1397773893]) : (!vm.ref<!hal.executable_cache>, i32 ...) -> i32
    %_simple_mul_ex_dispatch_0_binary_spirv = vm.const.ref.rodata @_simple_mul_ex_dispatch_0_binary_spirv : !vm.ref<!iree.byte_buffer>
    %ref_10 = vm.switch.ref %3[%_simple_mul_ex_dispatch_0_binary_spirv] else %null : !vm.ref<!iree.byte_buffer>
    %ref_11 = vm.call @hal.executable_cache.prepare(%ref_6, %_executable_layout_0_9, %c7, %ref_10) : (!vm.ref<!hal.executable_cache>, !vm.ref<!hal.executable_layout>, i32, !vm.ref<!iree.byte_buffer>) -> !vm.ref<!hal.executable>
```

Before we had serialized binary data for each target under each executable, so these switches always had non-null values. Now there's a separate executable for VMLA and for Vulkan.

Reply:
Here are IR dumps before and after, which I'm using to figure out where to make changes: https://gist.github.com/ScottTodd/907bcb753e5042db48466b3cf82a64d8

Reply:
Currently thinking about making changes around here: https://github.com/google/iree/blob/07412b25076ad6617f03c7bf6eb956b4466d40cb/iree/compiler/Dialect/HAL/Conversion/HALToVM/ConvertExecutableOps.cpp#L108-L121 

I could also see adding branching at a higher level, or making the underlying HAL ops tolerant to receiving empty data.

Also note that there is no graceful error handling for this new failure in hal_module.cc: https://github.com/google/iree/blob/07412b25076ad6617f03c7bf6eb956b4466d40cb/iree/modules/hal/hal_module.cc#L670-L680
(` executable_data->data` just asserts)

Reply:
The higher level to branch at is in MaterializeResourceCaches: https://github.com/google/iree/blob/8255d344b3d551fbac3348a8d454909b39a763ed/iree/compiler/Dialect/HAL/Transforms/MaterializeResourceCaches.cpp#L194-L255 by using `hal.device.switch`. `DeviceSwitchBuilder` should be used for this, but using it will require changing it back to take `OpBuilder` (after https://github.com/google/iree/pull/3156)

Reply:
Added `hal.device.switch` with cases for each target, but had to hack around those recent `DeviceSwitchBuilder` changes :/